pkg/provider/azure_vmss_cache.go

```diff
@@ -294,24 +313,36 @@ func (ss *ScaleSet) deleteCacheForNode(nodeName string) error {
 	return nil
 }
 
-func (ss *ScaleSet) newAvailabilitySetNodesCache() (*azcache.TimedCache, error) {
+func (ss *ScaleSet) newNonVmssUniformNodesCache() (*azcache.TimedCache, error) {
 	getter := func(key string) (interface{}, error) {
-		vmNames := sets.NewString()
+		klog.V(2).Infof("calling getter function of newNonVmssUniformNodesCache with key: %s", key)
+		vmssFlexVMNodeNames := sets.NewString()
+		vmssFlexVMProviderIDs := sets.NewString()
+		avSetVMNodeNames := sets.NewString()
+		avSetVMProviderIDs := sets.NewString()
 		resourceGroups, err := ss.GetResourceGroups()
 		if err != nil {
 			return nil, err
 		}
 
-		vmList := make([]compute.VirtualMachine, 0)
 		for _, resourceGroup := range resourceGroups.List() {
 			vms, err := ss.Cloud.ListVirtualMachines(resourceGroup)
 			if err != nil {
-				return nil, fmt.Errorf("newAvailabilitySetNodesCache: failed to list vms in the resource group %s: %w", resourceGroup, err)
+				return nil, fmt.Errorf("getter function of nonVmssUniformNodesCache: failed to list vms in the resource group %s: %w", resourceGroup, err)
 			}
 			for _, vm := range vms {
-				if vm.Name != nil {
-					vmNames.Insert(to.String(vm.Name))
-					vmList = append(vmList, vm)
+				if vm.OsProfile != nil && vm.OsProfile.ComputerName != nil {
+					if vm.VirtualMachineScaleSet != nil {
+						vmssFlexVMNodeNames.Insert(strings.ToLower(to.String(vm.OsProfile.ComputerName)))
+						if vm.ID != nil {
+							vmssFlexVMProviderIDs.Insert(ss.ProviderName() + "://" + to.String(vm.ID))
+						}
+					} else {
+						avSetVMNodeNames.Insert(strings.ToLower(to.String(vm.OsProfile.ComputerName)))
```

Comment:
it's changing avset cache logic already, originally it's `to.String(vm.Name)`, now it's `strings.ToLower(to.String(vm.OsProfile.ComputerName))`

do you have example what's the diff between these two values, and why use `strings.ToLower`

Reply:
After testing different types of VMs (Flex, Uniform, AvSet), I found out the naming conventions of nodeName. More specifically:
- No matter which node type it is, the nodeName is always equal to lower case of OS HostName (namely, strings.ToLower(to.String(vm.OsProfile.ComputerName))). This is because the kubelet report the nodeName using OS HostName when a new node just joined the cluster.
- When using AvSet VM, the OS hostname is always the same as VM name. This is why using vm name for avSet vm also works.

So, to unify them, we should always use OS HostName.